content/posts/lets-code-contact-form-in-phoenix-part-two.md

```diff
@@ -0,0 +1,240 @@
+
+```elixir
+# lib/my_app/support.ex
+
+defmodule MyApp.Support do
+  @moduledoc """
+  The Support module is the interface to customer support for the application.
+  Look to this module when you want to provide support to application users.
+  """
+
+  # Alias the `Support.Message` module for easy reference later in this module.
+  alias __MODULE__.Message
+
+  @doc """
+  Returns a changeset for a Support Message.
+  """
+  @spec change_message() :: Ecto.Changeset.t()
+  def change_message do
+    %Message{}
+    |> Message.changeset(%{})
+  end
+end
+```
+
+This very simple module and `change_message` function is enough to make the test pass. This function can be used to build a changeset for a new message. Now I will add tests for sending a message.
+
+```elixir
+# test/my_app/support_test.exs
+
+defmodule MyApp.SupportTest do
+  describe "sending a message" do
+    test "with valid fields" do
+      fields = %{
+        email: "barry@bluejeans.test",
+        subject: "Halp Me",
+        body: "I am having problems."
+      }
+
+      assert {:ok, %Support.Message{} = message} = Support.send_message(fields)
+      assert message.email == "barry@bluejeans.test"
+      assert message.subject == "Halp Me"
+      assert message.body == "I am having problems."
+    end
+
+    test "with invalid fields" do
+      assert {:error, cset} = Support.send_message(%{})
+
+      refute cset.valid?
+      assert "can't be blank" in errors_on(cset).email
+    end
+  end
+end
+```
+
+Here, I added a new `describe` block for testing the functionality of sending a message. This block has two tests: one for a successful message send and one for a failed message send. Successfully sending a message results in an `:ok` tuple response with a sent message. Failing to send a message returns an `:error` tuple with a changeset.
+
+Running the tests results in the following failure:
+
+```
+** (UndefinedFunctionError) function LearnU.Support.send_message/1 is undefined or private
+```
+
+I'm missing the `send_message` function. I can open up the `MyApp.Support` module again and add the following definition of the function to the bottom:
+
+```elixir
+# lib/my_app/support.ex
+
+defmodule MyApp.Support do
+  @doc """
+  Sends a message with given fields to the support team. Fields should include:
+  - `email`: Email for the user submitting the message for the support team to respond.
+  - `subject`: A short, general title as to the reason for support
+  - `body`: A longer description as to the reason for support and how the
+            support team might be able to help.
+
+  If the message is invalid, an `{:error, changeset}` is returned. Returns
+  `{:ok, message}` on success.
+  """
+  @spec send_message(map) :: {:ok, Message.t()} | {:error, Ecto.Changeset.t()}
+  def send_message(fields) when is_map(fields) do
```

Comment:
Is it worth commenting on why you wanted to add the guard clause here? If someone were following along, would they want additional functions that handle when `fields` isn't a map (or would they let it crash)?